### ISSUE TITLE
ERROR: Failed to autoscale: HTTPSConnectionPool(host='172.30.0.1', port=443): Read timed out. (read timeout=10)

### ISSUE THREAD

**Issue by cristian98149** [caas-team/py-kube-downscaler#118]
## Issue

During normal execution, after 2 days and 20 minutes after scaling up all workloads, the following error is returned:
```
2024-11-25 06:21:17,071 ERROR: Failed to autoscale: HTTPSConnectionPool(host='172.30.0.1', port=443): Read timed out. (read timeout=10)
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 466, in _make_request
    self._validate_conn(conn)
  File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 1095, in _validate_conn
    conn.connect()
  File "/usr/local/lib/python3.10/site-packages/urllib3/connection.py", line 730, in connect
    sock_and_verified = _ssl_wrap_socket_and_match_hostname(
  File "/usr/local/lib/python3.10/site-packages/urllib3/connection.py", line 909, in _ssl_wrap_socket_and_match_hostname
    ssl_sock = ssl_wrap_socket(
  File "/usr/local/lib/python3.10/site-packages/urllib3/util/ssl_.py", line 469, in ssl_wrap_socket
    ssl_sock = _ssl_wrap_socket_impl(sock, context, tls_in_tls, server_hostname)
  File "/usr/local/lib/python3.10/site-packages/urllib3/util/ssl_.py", line 513, in _ssl_wrap_socket_impl
    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/local/lib/python3.10/ssl.py", line 513, in wrap_socket
    return self.sslsocket_class._create(
  File "/usr/local/lib/python3.10/ssl.py", line 1104, in _create
    self.do_handshake()
  File "/usr/local/lib/python3.10/ssl.py", line 1375, in do_handshake
    self._sslobj.do_handshake()
TimeoutError: _ssl.c:990: The handshake operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/requests/adapters.py", line 667, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 843, in urlopen
    retries = retries.increment(
  File "/usr/local/lib/python3.10/site-packages/urllib3/util/retry.py", line 474, in increment
    raise reraise(type(error), error, _stacktrace)
  File "/usr/local/lib/python3.10/site-packages/urllib3/util/util.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 789, in urlopen
    response = self._make_request(
  File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 490, in _make_request
    raise new_e
  File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 468, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=conn.timeout)
  File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 369, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='172.30.0.1', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/kube_downscaler/main.py", line 86, in run_loop
    scale(
  File "/kube_downscaler/scaler.py", line 1362, in scale
    forced_uptime = pods_force_uptime(api, namespaces)
  File "/kube_downscaler/scaler.py", line 203, in pods_force_uptime
    for pod in pods:
  File "/usr/local/lib/python3.10/site-packages/pykube/query.py", line 197, in __iter__
    return iter(self.query_cache["objects"])
  File "/usr/local/lib/python3.10/site-packages/pykube/query.py", line 187, in query_cache
    cache["response"] = self.execute().json()
  File "/usr/local/lib/python3.10/site-packages/pykube/query.py", line 161, in execute
    r = self.api.get(**kwargs)
  File "/usr/local/lib/python3.10/site-packages/pykube/http.py", line 465, in get
    return self.session.get(*args, **self.get_kwargs(**kwargs))
  File "/usr/local/lib/python3.10/site-packages/requests/sessions.py", line 602, in get
    return self.request("GET", url, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.10/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/pykube/http.py", line 181, in send
    response = self._do_send(request, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/requests/adapters.py", line 713, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='172.30.0.1', port=443): Read timed out. (read timeout=10)
```

Container version is: 24.10.1

What the error means?

Also, the POD where py-kube-downscaler shound't have a failins status ("CrashLoopbackOff" for example)?

**Comment by samuel-esp** [comment]
Hi @cristian98149 thank you for reporting this error which seems to be the same as #105. If possible, could you provide details about the resources the tool was attempting to downscale or upscale when this error occurred, as well as its behavior afterward? (an extensive log would be perfect).

As i mentioned earlier this issue was already reported, it shouldn't be causing downtime to the tool and in the only case reported, the user told me that at the end the Kubedownscaler was still able to do its work. My take is that it could be caused by 

1) A connection request is placed on hold by the Kubernetes API Server when it exceeds the limit of `x` concurrent operations. If there is no communication between the Client (KubeDownscaler) and the Server (Kubernetes API Server), the held connection request will automatically close itself after a fixed amount of time (that is 10 seconds by default for our tool).
2) Network performances

Unfortunately I still didn't manage to replicate this issue, once this PR #117 is merged you should be able to set a custom timeout inside the Kube Downscaler Args (`--api-server-timeout`) in order to avoid or mitigate this issue. It will be available in the next release by the end of the month

**Comment by cristian98149** [comment]
Hi @samuel-esp ,

I confirm that the process is not impacted by this error as it continues to scale workloads.

I think the error is not caused by the scaling of a workload as it happened 20 mins after last workload scaling up and further today it didn't happen. So I think the error isn't related to a scaling action.
